core/src/main/java/com/exadel/etoolbox/linkinspector/core/servlets/SettingsDataSourceServlet.java

```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    protected void doGet(final SlingHttpServletRequest request, final SlingHttpServletResponse response) {
+
+        String rootPath = request.getRequestPathInfo().getResourcePath();
+        Resource mainTab = request.getResourceResolver().getResource(rootPath + "/main");
+        Resource advancedTab = request.getResourceResolver().getResource(rootPath + "/advanced");
+
+        List<Resource> tabs = new ArrayList<>();
+        tabs.add(mainTab);
+        tabs.add(advancedTab);
+
+        Map<String, List<AttributeDefinition>> serviceSettings = getServiceSettings();
+        for (Map.Entry<String, List<AttributeDefinition>> stringListEntry : serviceSettings.entrySet()) {
+            String serviceName = stringListEntry.getKey();
+            List<Resource> innerFields = new ArrayList<>();
+            for (AttributeDefinition attributeDefinition : stringListEntry.getValue()) {
+                Map<String, Object> fieldProperties = new HashMap<>();
+                fieldProperties.put(JcrResourceConstants.SLING_RESOURCE_TYPE_PROPERTY, getResourceTypeByType(attributeDefinition.getType()));
+
+                fieldProperties.put("name", "./" + serviceName + "." + attributeDefinition.getID());
+                fieldProperties.put("fieldLabel", attributeDefinition.getName());
+                fieldProperties.put("fieldDescription", attributeDefinition.getDescription());
+                Resource textField = ServletUtil.createResource(request.getResourceResolver(), serviceName + "." + attributeDefinition.getID(), fieldProperties, Collections.emptyList());
+                innerFields.add(textField);
+            }
+            Resource tab = ServletUtil.createTab(request.getResourceResolver(), serviceName, splitCamelCase(StringUtils.substringAfterLast(serviceName, ".")), innerFields);
+            tabs.add(tab);
+        }
+
+        DataSource dataSource = new SimpleDataSource(tabs.iterator());
+        request.setAttribute(DataSource.class.getName(), dataSource);
+    }
+
+    private String splitCamelCase(String s) {
+        return s.replaceAll(
+                String.format("%s|%s|%s",
+                        "(?<=[A-Z])(?=[A-Z][a-z])",
+                        "(?<=[^A-Z])(?=[A-Z])",
+                        "(?<=[A-Za-z])(?=[^A-Za-z])"
+                ),
+                " "
+        );
+    }
+
+    private String getResourceTypeByType(int type) {
+        String result;
+        switch (type) {
+            case 3:
+                result = "granite/ui/components/coral/foundation/form/numberfield";
+                break;
+            default:
+                result = "granite/ui/components/coral/foundation/form/textfield";
+                break;
+        }
+        return result;
+    }
+
+    private Map<String, List<AttributeDefinition>> getServiceSettings() {
+        Map<String, List<AttributeDefinition>> result = new HashMap<>();
+        for (LinkResolver linkResolver : linkResolvers) {
```

Comment:
`linkResolvers` can be null so it is better to wrap in `CollectionUtils.emptyIfNull()`

Reply:
thanks, added NPE check